### DISCUSSION TITLE
Custom sentenciser splitting unnecessarily

### DISCUSSION THREAD

**Discussion by emmacooke123** [explosion/spaCy#7564] [Help: Coding & Implementations]
I am using spaCy pipelines in order to split sentences more specifically to what I require. I am achieving it this by creating a custom pipe in the pipeline which modifies token.is_sent_start : 
```
nlp = spacy.load("en_core_web_sm")
nlp.add_pipe("split_on_bracketed_letters_numbers", before="parser")
processed_text = nlp(text)
```
I have created a function "split_on_bracketed_letters_numbers" which splits sentences on bracketed letters and numbers such that a sentence like:
`text = "According to the sections (1) and (2) of this page."`
should become:
`processed_text = ["According to the sections", "(1) and (2) of this page."]`
However, I do not see this response the response I do see is: 
`processed_text = ["According to the sections", "(1) and", "(2) of this page."]`

In order to try and debug this, at the end of the function "split_on_bracketed_letters_numbers" I print out the token.is_sent_start for each token: 
```
for token in doc:
    print(token.text, token.is_sent_start)
```
And I get the print: 
```
According True
to None
the None
sections None
( True
1 None
) None
and False
( False
2 None
) None
of False
this None
page None
. None
```
Which indicates to me there should be two sentence starts here, and so two sentences. However I am seeing three sentences in my output. 
As I understand it, from the [documentation](https://spacy.io/usage/processing-pipelines#custom-components-simple)  any` token.is_sent_start `which is set to `True` or `False` should dominate any` token.is_sent_start `'s which are allocated by the parser. However, this is does not reflect the result I am seeing.

Am I misunderstanding something here? Or could this be a bug?

Here's a sample code to replicate the problem:
```
import re
import spacy
from spacy.language import Language

# example matches: (a), (b), (1), (12), (xii)
BRACKETED_CHAR_REGEX = r"\(([0-9]|[a-r]|v|x){1,3}\)"
# example matches: (a) and (b), (1) or (12), (xii) to (xx)
BRACKETED_CHAR_REGEX_WITH_AND_OR_TO = (
    r"\(([0-9]|[a-r]|v|x){1,3}\)(and|or|to)\(([0-9]|[a-r]|v|x){1,3}\)"
)


@Language.component("split_on_bracketed_letters_numbers")
def split_on_bracketed_letters_numbers(doc):
    """ Spacy pipeline to split sentences on bracketed letters or numbers. If the bracketed letter
    or number is followed by 'and' or 'or' then another bracketed letter or number, do not
    split the sentence. """
    tokens_since_new_sentence = 0
    for token in doc[:-4]:
        # note this loop is over tokens (words or punctuation), not characters
        tokens_since_new_sentence += 1
        if token.text == "(":
            # look for regex match in next 3 tokens eg. ["(", "a", ")"]
            # cases which several bracketed letters are still 3 tokens
            # eg. '(iii)' -> ['(', 'iii', ')']
            next_tokens = "".join(
                [next_token.text for next_token in doc[token.i : token.i + 3]]
            )
            previous_and_next_tokens = "".join(
                [next_token.text for next_token in doc[token.i - 4 : token.i + 3]]
            )
            if re.search(BRACKETED_CHAR_REGEX_WITH_AND_OR_TO, previous_and_next_tokens):
                # we also need to prevent a new sentence starting in the following token
                # which would otherwise lead to
                # "(a) This is a sentence." -> "(a)", "This is a sentence."
                doc[token.i + 3].is_sent_start = False
                doc[token.i].is_sent_start = False
                # print(doc[token.i])
            elif re.search(BRACKETED_CHAR_REGEX, next_tokens):
                # we have found a bracketed letter, so need to split the sentence
                # don't invoke this rule for a new sentence if a sentence already just started
                # this avoids breaking (1) (a) into two sentences
                if tokens_since_new_sentence > 3:
                    doc[token.i].is_sent_start = True
                # we also need to prevent a new sentence starting in the following token
                # which would otherwise lead to
                # "(a) This is a sentence." -> "(a)", "This is a sentence."
                doc[token.i + 3].is_sent_start = False
        if doc[token.i].is_sent_start:
            tokens_since_new_sentence = 0
    # Print out the is_sent_start before entering the parser
    for token in doc:
        print(token.text, doc[token.i].is_sent_start)
    return doc


nlp = spacy.load("en_core_web_sm")
# Add customer sentence splitting rules to standard pipeline
nlp.add_pipe("split_on_bracketed_letters_numbers", before="parser")
text = "According to the sections (1) and (2) of this page."
print("------------------------")
print("Input sentence:")
print(text)
print("------------------------")
print("Tokens in custom pipeline:")
print("------------------------")
processed_text = nlp(text)
print("------------------------")
print("Split sentences:")
print("------------------------")
for sentence in processed_text.sents:
    print(sentence.text)
print("------------------------")
print("Tokens after full pipeline:")
print("------------------------")
for token in processed_text:
    print(token.text, processed_text[token.i].is_sent_start)
```

**Comment by polm** [comment]
That does sound weird. I made some sample code with a toy component to set all words after the first to have `is_sent_start = False` and it worked the way I'd expect. 

```
import spacy

nlp = spacy.load("en_core_web_sm")

text = "I like ginger. I like books. I like spaCy."

for sent in nlp(text).sents:
    print(sent)

for token in nlp(text):
    print(token, token.is_sent_start, sep="\t")

from spacy.language import Language

@Language.component("antisentence")
def antisentence(doc):
    for tok in doc[1:]:
        tok.is_sent_start = False

    return doc

print("-----")
nlp.add_pipe("antisentence", before="parser")

for sent in nlp(text).sents:
    print(sent)

for token in nlp(text):
    print(token, token.is_sent_start, sep="\t")
```

Could you share a self-contained example I could run to reproduce the issue?

**Reply by emmacooke123** [reply]
Thanks for getting back to me. I have updated the original thread with a bit of code to replicate the problem

**Reply by polm** [reply]
Huh, that is weird. 

If you explicitly set all the `sent_start` properties, for example by setting all `None` to `False` at the end of your component, the parser will not change them.

If you explicitly set `sent_start` on the 9th token in your sentence (`2`) to `False`, then you get the two sentences you wanted. 

I need to look at the code in more detail, thanks for pointing this out.

**Reply by emmacooke123** [reply]
Hi @polm, just wondering if there was any update on this yet? Cheers.

**Reply by polm** [reply]
Sorry for the lack of updates on this. We discussed it in the team, and while it does seem to be a bug the fix wasn't immediately obvious. We are working on it.

**Reply by emmacooke123** [reply]
Hi @polm, would you be able to give some insight on how high up on your list of priorities this bug is? I'm sure you guys have lots of other things which also need your attention. I'm just trying to gauge if it might be worth trying to fix this bug myself - although I haven't even peeked at the codebase yet. Thanks in advance for your info.

**Reply by adrianeboyd** [reply]
Unfortunately this is a fairly low priority bug for us at this point. You are welcome to give it a shot, but if you're not familiar with cython it may be a difficult exercise in debugging. I think the best place to start would be the methods related to `sent_start` in the transition system, so `cannot_sent_start`, `is_sent_start`, and `set_sent_start` here: https://github.com/explosion/spaCy/blob/f68fc29130f913f36e0f9339a902cee5f1ed7179/spacy/pipeline/_parser_internals/_state.pxd

From the test case Paul mentioned, it sounds likely to be an off-by-one error somewhere in how it handles `False`/`-1` values.
